pkg/scheduler/internal/cache/cache.go

```diff
@@ -70,6 +70,7 @@ type schedulerCache struct {
 	// a map from pod key to podState.
 	podStates map[string]*podState
 	nodes     map[string]*nodeInfoListItem
+	csiNodes  map[string]*storagev1beta1.CSINode
```

Comment:
I like this approach better, but there is one important missing part. The scheduler takes a snapshot of its cache and works with that snapshot for scheduling a pod. The snapshot remains unchanged until a scheduling decision is made. This helps avoid race conditions between a scheduling cycle and the informer events that update the scheduler cache in parallel. I think a similar snapshotting mechanism is needed for CSINodes.

Reply:
What kind of races does the snapshotting avoid?  I don't expect CSINode to be updated frequently in the cluster.

Reply:
Imagine a function that checks certain things for CSINode object. The first check passes, the object is updated in the cache, the second check runs. Often times our logic is not designed to handle such scenarios.

Reply:
We don't currently cache anything related to CSINode except the API objects themselves and they are not modified.

Reply:
If they are never modified, why do we have an [`update` function](https://github.com/kubernetes/kubernetes/pull/80084/files#diff-f4a894ca5e905aa5f613269fc967fe2cR587) for them? Also deletion of an object from the map could potentially cause problems. For example, a function may check that an object is in the map once and then index the map later on without further checks. If the object is deleted from the map, the map will return nil and could potentially cause nil pointer dereference.

Reply:
The API objects themselves are updated, however what I meant is that we do not cache any extra state based on those objects, and we don't modify those API objects in-memory to store temporary state.

My understanding of the snapshot capability is so that you don't have corner cases where the scheduler logic reads the value multiple times and gets different values.  I don't think that is an issue for this predicate. We only read and use the CSINode value once per node per plugin.  The CSINode object is also not expected to be frequently updated. It only changes when a driver is installed or upgraded in the cluster.

If the value we use in the predicate changes while we're still processing the predicate, that's ok. It's a similar scenario where the value might change after we have already made a scheduling decision.

Reply:
> My understanding of the snapshot capability is so that you don't have corner cases where the scheduler logic reads the value multiple times and gets different values. I don't think that is an issue for this predicate. We only read and use the CSINode value once per node per plugin.

If that's the case, I agree that snapshotting CSINodes is not necessary.

Reply:
@bsalamat, @msau42: thanks for looking into this. I made some small adjustments and I think this PR is ready for review again.